pubsub/pubsub.go

```diff
@@ -427,67 +476,70 @@ func (s *Subscription) Receive(ctx context.Context) (_ *Message, err error) {
+					s.postReceiveBatchHook(len(msgs))
+				}
+				s.mu.Lock()
+				defer s.mu.Unlock()
+				if err != nil {
+					// Non-retryable error from ReceiveBatch -> permanent error.
+					s.err = err
+				} else if len(msgs) > 0 {
+					s.q = append(s.q, msgs...)
+					if s.throughputStart.IsZero() {
+						s.throughputStart = time.Now()
+					}
+				}
+				close(s.waitc)
+				s.waitc = nil
+			}()
+		}
 		if len(s.q) > 0 {
 			// At least one message is available. Return it.
 			m := s.q[0]
 			s.q = s.q[1:]
-			// TODO(jba): pre-fetch more messages if the queue gets too small.
+			s.throughputCount++
 			return m, nil
 		}
-		if s.waitc != nil {
-			// A call to ReceiveBatch is in flight. Wait for it.
-			// TODO(jba): support multiple calls in flight simultaneously.
-			waitc := s.waitc
-			s.mu.Unlock()
-			select {
-			case <-waitc:
-				s.mu.Lock()
-				continue
-			case <-ctx.Done():
-				s.mu.Lock()
-				return nil, ctx.Err()
-			}
+		// No messages are available.
+		if s.throughputEnd.IsZero() && !s.throughputStart.IsZero() {
+			s.throughputEnd = time.Now()
 		}
-		// No messages are available and there are no calls to ReceiveBatch in flight.
-		// Make a call.
-		batchSize := s.updateBatchSize()
-		s.waitc = make(chan struct{})
+		// A call to ReceiveBatch must be in flight. Wait for it.
+		waitc := s.waitc
 		s.mu.Unlock()
-		// Even though the mutex is unlocked, only one goroutine can be here.
-		// The only way here is if s.waitc was nil. This goroutine just set
-		// s.waitc to non-nil while holding the lock.
-		if s.preReceiveBatchHook != nil {
-			s.preReceiveBatchHook(batchSize)
-		}
-		msgs, err := s.getNextBatch(ctx, batchSize)
-		if s.postReceiveBatchHook != nil {
-			s.postReceiveBatchHook(len(msgs))
-		}
-		s.mu.Lock()
-		// Update these even if there was an error; the next call to Receive will
-		// try again to get more messages, and updateBatchSize based on 0 messages
-		// proceesed, which is fine.
-		s.lastBatchRecv = time.Now()
-		s.lastBatchNumMsgs = len(msgs)
-		close(s.waitc)
-		s.waitc = nil
-		if err != nil {
-			// This goroutine's call failed, perhaps because its context was done.
-			// Some waiting goroutine will wake up when s.waitc is closed,
-			// go to the top of the loop, and (since s.q is empty and s.waitc is
-			// now nil) will try the RPC for itself.
-			return nil, err
+		select {
+		case <-waitc:
+			s.mu.Lock()
+			// Continue to top of loop.
+		case <-ctx.Done():
+			s.mu.Lock()
+			return nil, ctx.Err()
 		}
-		s.q = append(s.q, msgs...)
 	}
 }
 
 // getNextBatch gets the next batch of messages from the server and returns it.
-func (s *Subscription) getNextBatch(ctx context.Context, nMessages int) ([]*Message, error) {
+func (s *Subscription) getNextBatch(nMessages int) ([]*Message, error) {
 	var msgs []*driver.Message
-	err := retry.Call(ctx, gax.Backoff{}, s.driver.IsRetryable, func() error {
+	err := retry.Call(s.backgroundCtx, gax.Backoff{}, s.driver.IsRetryable, func() error {
```

Comment:
I think that if the user cancels the context they passed to Receive, then this call should be canceled immediately. IIUC the way you've got it, that won't happen until the user calls Shutdown. That is the right behavior for acks, but not ReceiveBatch.

Reply:
Why do you think it should be canceled? The `driver.ReceiveBatch` calls is no longer directly associated with the user's call to `Receive`; it's quite likely that we'll initiate a `driver.ReceiveBatch` call in the background but still return a message from the queue to the user. We wouldn't want to cancel the background `ReceiveBatch` just because the user canceled their `ctx` after `Receive` returned.

Reply:
OK, I guess that makes sense.